ambry-quota/src/main/java/com/github/ambry/quota/QuotaResourceSynchronizer.java

```diff
@@ -0,0 +1,102 @@
+  private final QuotaMetrics quotaMetrics;
+
+  /**
+   * Constructor for {@link QuotaResourceSynchronizer}.
+   * @param quotaMetrics {@link QuotaMetrics} object.
+   */
+  public QuotaResourceSynchronizer(QuotaMetrics quotaMetrics) {
+    this.quotaMetrics = quotaMetrics;
+  }
+
+  /**
+   * Acquires the lock for the quota resource represented by the specified {@link QuotaResource}.
+   * @param quotaResource {@link QuotaResource} object representing the quota resource for which lock needs to be
+   *                                           acquired.
+   */
+  public void lock(QuotaResource quotaResource) {
+    quotaResourceMutexCache.get(quotaResource).lock();
+  }
+
+  /**
+   * Releases the lock for the quota resource represented by the specified {@link QuotaResource}.
+   * @param quotaResource {@link QuotaResource} object representing the quota resource for which lock needs to be
+   *                      released.
+   */
+  public void unlock(QuotaResource quotaResource) {
+    ReentrantLock lock = quotaResourceMutexCache.mutexCache.getOrDefault(quotaResource, null);
+    if (lock != null) {
+      lock.unlock();
+    } else {
+      LOGGER.warn("Attempted unlock for {} when lock wasn't held.", quotaResource);
+      quotaMetrics.quotaResourceInvalidUnlockAttemptCount.inc();
+    }
+  }
+
+  /**
+   * A cache of unique mutexes for all {@link QuotaResource} objects that represent the same quota resource.
+   */
+  static class QuotaResourceMutexCache {
+    /*
+     * There are multiple optimization opportunities here.
+     * 1. If we never clean the cache, we don't really need all the complexity, and hence latency, of a ConcurrentHashMap.
+     *    We could just use an array of mutexes to synchronize. But doing that will add extra complexity (extra code,
+     *    extra tests etc). So unless there is a performance issue, using ConcurrentHashMap should be just fine.
+     * 2. This cache never removes entries until the service (Ambry frontend) is restarted. In the worst case the size
+     *    of the cache might grow upto the total number of QuotaResources (accounts + containers + services) for which quota
+     *    is being enforced. Periodically cleaning up cache is one way to reduce the memory pressure, but will add more
+     *    complexity to maintain thread safety (and maybe more latency to access the cache too). Ambry's frontend keeps
+     *    the accounts and containers in memory at multiple places currently (e.g, AccountService implementation), with the
+     *    assumption that the total number of accounts and containers are small enough to keep in memory. As long as that
+     *    assumption holds, not cleaning up this cache is fine. If that assumption doesn't hold, a lot of code might need
+     *    restructuring, and this cache cleanup should be taken up as part of that effort.
+     */
+    protected final ConcurrentHashMap<QuotaResource, ReentrantLock> mutexCache = new ConcurrentHashMap<>();
```

Comment:
ReentrantLock lets the same thread lock the same object more than once. Is there any case where the same thread can serve to multiple requests for the same QuotaResource?

Reply:
The reason why I didn't consider anything other than `ReentrantLock` is because that's the default `Lock` implementation in java standard library (The other implementation is `ReadWriteLock` which doesn't apply to this case). Even the `sycnhronized` blocks in java are re-entrant. Although there shouldn't be any case where a thread can serve multiple requests for the same `QuotaResource` at the same time, `ReentrantLocks` generally also help with taking the same locks in other blocks (or methods) too. Otherwise care should be taken that a lock once held in a flow, isn't required to be held again, or else the thread will block itself.